v2/internal/pkg/batch/concurrent_chan_worker.go

```diff
@@ -0,0 +1,383 @@
+package batch
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/openshift/oc-mirror/v2/internal/pkg/api/v2alpha1"
+	clog "github.com/openshift/oc-mirror/v2/internal/pkg/log"
+	"github.com/openshift/oc-mirror/v2/internal/pkg/mirror"
+	"github.com/openshift/oc-mirror/v2/internal/pkg/spinners"
+	"github.com/vbauerster/mpb/v8"
+	"github.com/vbauerster/mpb/v8/decor"
+)
+
+var (
+	copiedImages v2alpha1.CollectorSchema
+
+	errMsgHeader = "%ssome errors occurred during the mirroring"
+	errMsg       = errMsgHeader + ".\n" +
+		"\t Please review %s/%s for a list of mirroring errors.\n" +
+		"\t You may consider:\n" +
+		"\t * removing images or operators that cause the error from the image set config, and retrying\n" +
+		"\t * keeping the image set config (images are mandatory for you), and retrying\n" +
+		"\t * mirroring the failing images manually, if retries also fail."
+)
+
+type ChannelConcurrentBatch struct {
+	Log           clog.PluggableLoggerInterface
+	LogsDir       string
+	Mirror        mirror.MirrorInterface
+	MaxGoroutines int
+}
+
+type GoroutineResult struct {
+	err     *mirrorErrorSchema
+	imgType v2alpha1.ImageType
+	img     v2alpha1.CopyImageSchema
+}
+
+// Worker - the main batch processor
+func (o *ChannelConcurrentBatch) Worker(ctx context.Context, collectorSchema v2alpha1.CollectorSchema, opts mirror.CopyOptions) (v2alpha1.CollectorSchema, error) {
+	startTime := time.Now()
+
+	copiedImages = v2alpha1.CollectorSchema{
+		AllImages: []v2alpha1.CopyImageSchema{},
+	}
+
+	var errArray []mirrorErrorSchema
+	//OCPBUGS-44448
+	//TODO this bug fix will work only for images origin by digest only
+	//to cover scenarios where one digest is tagged by several images it is necessary to add a field on
+	//CopyImageSchema to store the digest number and rely on this field instead of img.Origin
+	alreadyDeleted := make(onlyKeyMap)
+
+	var m sync.RWMutex
+	var wg sync.WaitGroup
+
+	var mirrorMsg string
+	switch {
+	case opts.IsCopy():
+		mirrorMsg = "copying"
+	case opts.IsDelete():
+		mirrorMsg = "deleting"
+	}
+
+	opts.PreserveDigests = true
+
+	o.Log.Info("🚀 Start " + mirrorMsg + " the images...")
+	o.Log.Info("\U0001F4CC images to %s %d ", opts.Function, len(collectorSchema.AllImages))
+
+	total := len(collectorSchema.AllImages)
+	p := mpb.New()
+	results := make(chan GoroutineResult, total)
+	progressCh := make(chan int, total)
+	semaphore := make(chan struct{}, o.MaxGoroutines)
+	cancelCtx, cancel := opts.Global.CommandTimeoutContext()
+	defer cancel()
+
+	go func() {
+		defer close(results)
+		defer close(semaphore)
+
+		for _, img := range collectorSchema.AllImages {
+
+			select {
+			case <-cancelCtx.Done():
+				wg.Wait()
+				return
+			default:
+			}
+
+			sp := newSpinner(img, opts.LocalStorageFQDN, p)
+
+			semaphore <- struct{}{}
```

Comment:
@aguidirh - in memory semaphore - brilliant. One per go routine

Reply:
Yes, this will allow us to trigger only the number of routines defined in the flag `parallel-images`.

The semaphore will only trigger a new goroutine when there is a slot available in the channel buffer.